# ISLR Sec 5-4 Problem 5

In [1]:
from __future__ import print_function
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
import matplotlib
matplotlib.style.use('ggplot')
%matplotlib inline

In [2]:
df = pd.read_csv('../Data/Default.csv')
df.head(3)

Unnamed: 0 default student      balance        income
0           1      No      No   729.526495  44361.625074
1           2      No     Yes   817.180407  12106.134700
2           3      No      No  1073.549164  31767.138947

### (a)

In [3]:
LgR = LogisticRegression(multi_class='multinomial',solver='lbfgs')

### (b)

In [4]:
rand_state = 2

data = df[['income','balance']].values
target = df.default
x_train,x_test,y_train,y_test = train_test_split(data,target,random_state=rand_state)
LgR.fit(x_train,y_train)
y_pred = LgR.predict(x_test)
error_rate = 1 - accuracy_score(y_test,y_pred)
print(error_rate)

0.0272


### (c)

* Random state 0 split yeilds 0.0388 error rate.
* Random state 1 split yeilds 0.0316 error rate.
* Random state 2 split yeilds 0.0272 error rate.

The estimated test error rate varies depending on how the data is split.

### (d)

In [5]:
# one-hot-encoding of student column
df1 = pd.get_dummies(df,columns=['student'])
df1.head(3)

Unnamed: 0 default      balance        income  student_No  student_Yes
0           1      No   729.526495  44361.625074           1            0
1           2      No   817.180407  12106.134700           0            1
2           3      No  1073.549164  31767.138947           1            0

In [6]:
rand_state = 0

data1 = df1[['income','balance','student_No','student_Yes']].values
target1 = df1.default
x_train,x_test,y_train,y_test = train_test_split(data1,target1,random_state=rand_state)
LgR.fit(x_train,y_train)
y_pred = LgR.predict(x_test)
error_rate = 1 - accuracy_score(y_test,y_pred)
print(error_rate)

0.0296


Doesn't seem student feature improves prediction

In [7]:
# mean test error over 100 random splits without student feature
error_rate = np.zeros(100)
for k in range(100):
    x_train,x_test,y_train,y_test = train_test_split(data,target,random_state=rand_state)
    LgR.fit(x_train,y_train)
    y_pred = LgR.predict(x_test)
    error_rate[k] = 1 - accuracy_score(y_test,y_pred)
print('without student feature: mean error rate %0.3f' % error_rate.mean())

# mean test error rate over 100 random splits with student feature
error_rate = np.zeros(100)
for k in range(100):
    x_train,x_test,y_train,y_test = train_test_split(data1,target1,random_state=rand_state)
    LgR.fit(x_train,y_train)
    y_pred = LgR.predict(x_test)
    error_rate[k] = 1 - accuracy_score(y_test,y_pred)
print('with student feature: mean error rate %0.3f' % error_rate.mean())

without student feature: mean error rate 0.039
with student feature: mean error rate 0.030


There appears to be some improvement in average performance. Not entirely clear if this improvement is statistically significant. Bootstrapping (covered later) may provide a better answer.